File: mapbox_maps_flutter_mobile/ios/mapbox_maps_flutter/Sources/mapbox_maps_flutter/Classes/AnimationController.swift
import Foundation
@_spi(Experimental) import MapboxMaps
import Flutter

final class AnimationController: _AnimationManager {
    private static let errorCode = "0"

    func easeTo(cameraOptions: CameraOptions, mapAnimationOptions: MapAnimationOptions?) throws {
        var cameraDuration = 1.0
        if mapAnimationOptions != nil && mapAnimationOptions!.duration != nil {
            cameraDuration = Double(mapAnimationOptions!.duration!) / 1000.0
        }
        cancelable = camera.ease(to: cameraOptions.toCameraOptions(), duration: cameraDuration)
    }

    func flyTo(cameraOptions: CameraOptions, mapAnimationOptions: MapAnimationOptions?) throws {
        var cameraDuration = 1.0
        if mapAnimationOptions != nil && mapAnimationOptions!.duration != nil {
            cameraDuration = Double(mapAnimationOptions!.duration!) / 1000.0
        }
        cancelable = camera.fly(to: cameraOptions.toCameraOptions(), duration: cameraDuration)
    }

    func pitchBy(pitch: Double, mapAnimationOptions: MapAnimationOptions?) throws {
        throw FlutterError(code: AnimationController.errorCode, message: "Not available.", details: nil)
    }

    func scaleBy(amount: Double, screenCoordinate: ScreenCoordinate?, mapAnimationOptions: MapAnimationOptions?) throws {
        throw FlutterError(code: AnimationController.errorCode, message: "Not available.", details: nil)
    }

    func moveBy(screenCoordinate: ScreenCoordinate, mapAnimationOptions: MapAnimationOptions?) throws {
        throw FlutterError(code: AnimationController.errorCode, message: "Not available.", details: nil)
    }

    func rotateBy(first: ScreenCoordinate, second: ScreenCoordinate, mapAnimationOptions: MapAnimationOptions?) throws {
        throw FlutterError(code: AnimationController.errorCode, message: "Not available.", details: nil)
    }

    func cancelCameraAnimation() throws {
        if cancelable != nil {
            cancelable?.cancel()
        }
    }

    private var camera: CameraAnimationsManager
    private var cancelable: MapboxMaps.Cancelable?
    init(withMapView mapView: MapView) {
        self.camera = mapView.camera
    }
}

File: mapbox_maps_flutter_mobile/android/src/main/kotlin/com/mapbox/maps/mapbox_maps/AnimationController.kt
package com.mapbox.maps.mapbox_maps

import android.content.Context
import com.mapbox.common.Cancelable
import com.mapbox.maps.MapboxMap
import com.mapbox.maps.mapbox_maps.pigeons.*
import com.mapbox.maps.plugin.animation.easeTo
import com.mapbox.maps.plugin.animation.flyTo
import com.mapbox.maps.plugin.animation.moveBy
import com.mapbox.maps.plugin.animation.pitchBy
import com.mapbox.maps.plugin.animation.rotateBy
import com.mapbox.maps.plugin.animation.scaleBy

class AnimationController(private val mapboxMap: MapboxMap, private val context: Context) : _AnimationManager {
  var cancelable: Cancelable? = null
  override fun easeTo(
    cameraOptions: CameraOptions,
    mapAnimationOptions: MapAnimationOptions?
  ) {
    cancelable = mapboxMap.easeTo(
      cameraOptions.toCameraOptions(context),
      mapAnimationOptions?.toMapAnimationOptions()
    )
  }

  override fun flyTo(
    cameraOptions: CameraOptions,
    mapAnimationOptions: MapAnimationOptions?
  ) {
    cancelable = mapboxMap.flyTo(
      cameraOptions.toCameraOptions(context),
      mapAnimationOptions?.toMapAnimationOptions()
    )
  }

  override fun moveBy(
    screenCoordinate: ScreenCoordinate,
    mapAnimationOptions: MapAnimationOptions?
  ) {
    cancelable = mapboxMap.moveBy(
      screenCoordinate.toScreenCoordinate(context),
      mapAnimationOptions?.toMapAnimationOptions()
    )
  }

  override fun rotateBy(
    first: ScreenCoordinate,
    second: ScreenCoordinate,
    mapAnimationOptions: MapAnimationOptions?
  ) {
    cancelable = mapboxMap.rotateBy(
      first.toScreenCoordinate(context),
      second.toScreenCoordinate(context),
      mapAnimationOptions?.toMapAnimationOptions()
    )
  }

  override fun scaleBy(
    amount: Double,
    screenCoordinate: ScreenCoordinate?,
    mapAnimationOptions: MapAnimationOptions?
  ) {
    cancelable = mapboxMap.scaleBy(
      amount,
      screenCoordinate?.toScreenCoordinate(context),
      mapAnimationOptions?.toMapAnimationOptions()
    )
  }

  override fun pitchBy(pitch: Double, mapAnimationOptions: MapAnimationOptions?) {
    cancelable = mapboxMap.pitchBy(pitch, mapAnimationOptions?.toMapAnimationOptions())
  }

  override fun cancelCameraAnimation() {
    cancelable?.cancel()
  }
}